opencv 4.5.1 build fails with numpy 2.3.0 (all from current develop branch)

## P4A opencv build fails from develop branch

### Background
- I am creating an android app (apk) with `opencv` & `numpy` from the `develop` branch.
- The current [numpy](https://github.com/kivy/python-for-android/blob/develop/pythonforandroid/recipes/numpy/__init__.py) version is `v2.3.0` & [opencv](https://github.com/kivy/python-for-android/blob/develop/pythonforandroid/recipes/opencv/__init__.py) version is `4.5.1`.

### Build error
```bash
[100%] Building CXX object modules/gapi/CMakeFiles/opencv_gapi.dir/src/backends/common/serialization.cpp.o
[100%] Building CXX object modules/gapi/CMakeFiles/opencv_gapi.dir/src/backends/streaming/gstreamingbackend.cpp.o
[100%] Building CXX object modules/gapi/CMakeFiles/opencv_gapi.dir/src/backends/ie/bindings_ie.cpp.o
[100%] Linking CXX shared library ../../lib/arm64-v8a/libopencv_gapi.so
[100%] Built target opencv_gapi
[100%] Building CXX object modules/python3/CMakeFiles/opencv_python3.dir/__/src2/cv2.cpp.o
/home/somnath/codes/git/my-org/vision-ai/onnxruntime/.buildozer/android/platform/build-arm64-v8a/build/other_builds/opencv/arm64-v8a__ndk_target_28/opencv/modules/python/src2/cv2.cpp:33:10: fatal error: 'numpy/ndarrayobject.h' file not found
#include <numpy/ndarrayobject.h>
         ^~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[3]: *** [modules/python3/CMakeFiles/opencv_python3.dir/build.make:76: modules/python3/CMakeFiles/opencv_python3.dir/__/src2/cv2.cpp.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:2306: modules/python3/CMakeFiles/opencv_python3.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2313: modules/python3/CMakeFiles/opencv_python3.dir/rule] Error 2
make: *** [Makefile:680: opencv_python3] Error 2
```

### My observations
- The above error doesn't occur when we use `numpy` versions `1.26.5` with `opencv 4.5.1`. Here is an example of my existing app's [numpy recipe](https://github.com/daslearning-org/image-to-animation-offline/blob/main/kivy/p4a_local_recipes/numpy/__init__.py) & [opencv recipe](https://github.com/daslearning-org/image-to-animation-offline/blob/main/kivy/p4a_local_recipes/opencv/__init__.py).

- OpenCV with `NDK >= 26`. When we use NDK > 25, we get below error with `-version` argument, this error is not coming when using `NDK 25b`
```bash
[100%] Linking CXX shared module ../../lib/arm64-v8a/python3/cv2.cpython-311-x86_64-linux-gnu.so
clang++: error: unknown argument '-version'; did you mean '--version'?
make[3]: *** [modules/python3/CMakeFiles/opencv_python3.dir/build.make:112: lib/arm64-v8a/python3/cv2.cpython-311-x86_64-linux-gnu.so] Error 1
make[2]: *** [CMakeFiles/Makefile2:2306: modules/python3/CMakeFiles/opencv_python3.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2313: modules/python3/CMakeFiles/opencv_python3.dir/rule] Error 2
make: *** [Makefile:680: opencv_python3] Error 2
```


#### My environment details
- My development python version `3.11.13` 
- Buildozer version `1.5.0`
- P4A branch `develop` which uses `hostpython3` & `python3` at `3.11.5`

> Please let me know if you require any further details. Thank you

--- a/pythonforandroid/recipes/opencv/__init__.py
+++ b/pythonforandroid/recipes/opencv/__init__.py
@@ -15,7 +15,7 @@ class OpenCVRecipe(NDKRecipe):
         build of most of the libraries of the opencv's package, so we can
         process images, videos, objects, photos...
     '''
-    version = '4.5.1'
+    version = '4.12.0'
     url = 'https://github.com/opencv/opencv/archive/{version}.zip'
     depends = ['numpy']
     patches = ['patches/p4a_build.patch']
@@ -68,8 +68,9 @@ class OpenCVRecipe(NDKRecipe):
             python_link_version = self.ctx.python_recipe.link_version
             python_library = join(python_link_root,
                                   'libpython{}.so'.format(python_link_version))
-            python_include_numpy = join(python_site_packages,
-                                        'numpy', 'core', 'include')
+            python_include_numpy = join(
+                self.ctx.get_python_install_dir(arch.arch), "numpy/_core/include",
+            )
 
             shprint(sh.cmake,
                     '-DP4A=ON',
@@ -136,6 +137,15 @@ class OpenCVRecipe(NDKRecipe):
 
                     self.get_build_dir(arch.arch),
                     _env=env)
+
+            # patch link.txt for unsupported flag
+            link_txt = 'modules/python3/CMakeFiles/opencv_python3.dir/link.txt'
+            with open(link_txt, 'r+') as f:
+                content = f.read().replace('-version', ' ')
+                f.seek(0)
+                f.write(content)
+                f.truncate()
+
             shprint(sh.make, '-j' + str(cpu_count()), 'opencv_python' + python_major)
             # Install python bindings (cv2.so)
             shprint(sh.cmake, '-DCOMPONENT=python', '-P', './cmake_install.cmake')

--- a/pythonforandroid/recipes/opencv/patches/p4a_build.patch
+++ b/pythonforandroid/recipes/opencv/patches/p4a_build.patch
@@ -1,31 +1,22 @@
-This patch allow that the opencv's build command correctly detects our version
-of python, so we can successfully build the python bindings (cv2.so)
---- opencv-4.0.1/cmake/OpenCVDetectPython.cmake.orig	2018-12-22 08:03:30.000000000 +0100
-+++ opencv-4.0.1/cmake/OpenCVDetectPython.cmake	2019-01-31 11:33:10.896502978 +0100
-@@ -175,7 +175,7 @@ if(NOT ${found})
+diff '--color=auto' -uNr opencv-4.12.0/cmake/OpenCVDetectPython.cmake opencv-4.12.0.mod/cmake/OpenCVDetectPython.cmake
+--- opencv-4.12.0/cmake/OpenCVDetectPython.cmake	2025-07-02 13:24:13.000000000 +0530
++++ opencv-4.12.0.mod/cmake/OpenCVDetectPython.cmake	2025-09-20 22:22:14.961944470 +0530
+@@ -175,7 +175,7 @@
        endif()
      endif()
-
--    if(NOT ANDROID AND NOT IOS)
-+    if(P4A OR NOT ANDROID AND NOT IOS)
+ 
+-    if(NOT ANDROID AND NOT IOS AND NOT XROS)
++    if(P4A OR NOT ANDROID AND NOT IOS AND NOT XROS)
        if(CMAKE_HOST_UNIX)
-         execute_process(COMMAND ${_executable} -c "from distutils.sysconfig import *; print(get_python_lib())"
+         execute_process(COMMAND ${_executable} -c "from sysconfig import *; print(get_path('purelib'))"
                          RESULT_VARIABLE _cvpy_process
-@@ -244,7 +244,7 @@ if(NOT ${found})
-                           OUTPUT_STRIP_TRAILING_WHITESPACE)
-         endif()
-       endif()
--    endif(NOT ANDROID AND NOT IOS)
-+    endif(P4A OR NOT ANDROID AND NOT IOS)
-   endif()
-
-   # Export return values
---- opencv-4.0.1/modules/python/CMakeLists.txt.orig	2018-12-22 08:03:30.000000000 +0100
-+++ opencv-4.0.1/modules/python/CMakeLists.txt	2019-01-31 11:47:17.100494908 +0100
+diff '--color=auto' -uNr opencv-4.12.0/modules/python/CMakeLists.txt opencv-4.12.0.mod/modules/python/CMakeLists.txt
+--- opencv-4.12.0/modules/python/CMakeLists.txt	2025-07-02 13:24:13.000000000 +0530
++++ opencv-4.12.0.mod/modules/python/CMakeLists.txt	2025-09-20 22:23:15.124356524 +0530
 @@ -3,7 +3,7 @@
  # ----------------------------------------------------------------------------
  if(DEFINED OPENCV_INITIAL_PASS)  # OpenCV build
-
+ 
 -if(ANDROID OR APPLE_FRAMEWORK OR WINRT)
 +if(ANDROID AND NOT P4A OR APPLE_FRAMEWORK OR WINRT)
    ocv_module_disable_(python2)

